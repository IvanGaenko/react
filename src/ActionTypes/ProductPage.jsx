export const GET_SINGLE_ITEM = 'GET_SINGLE_ITEM';
export const GET_NEXT_ITEM = 'GET_NEXT_ITEM';
export const GET_PREV_ITEM = 'GET_PREV_ITEM';
