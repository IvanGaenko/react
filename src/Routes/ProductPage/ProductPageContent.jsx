import React, { Component } from "react";
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';

const styles = () => (
  {
    card: {
      minWidth: 700,
      maxWidth: 1000,
      marginRight: 40,
      marginLeft: 40,
      marginBottom:30
    },
    media: {
      height: 800,
      width: 1000,
      paddingTop: '56.25%', // 16:9
    },
    button: {
      minWidth: '100%',
    },
  }
);

class ProductPageContent extends Component {

  render() {
    const { classes, currencyPrice, addToCart } = this.props;
    const { id, title, author, image, year } = this.props.singleProduct;
      return (
        <Card className={classes.card}>
          <CardMedia
            className={classes.media}
            image={image}
            title={id}
          />
          <CardContent>
            <Typography gutterBottom variant="headline" component="h2">
              {title}
            </Typography>
            <Typography gutterBottom variant="headline" component="h2">
              {author}
            </Typography>
            <Typography component="p">
              Year: {year}
            </Typography>
            <Typography  component="p">
              Price: {currencyPrice}
            </Typography>
          </CardContent>

          <CardActions>
            <Button
              id="add-cart"
              className={classes.button}
              size="small" color="primary"
              onClick={addToCart}
            >
              Buy
            </Button>
          </CardActions>
        </Card>
    )
  }
};

export default withStyles(styles)(ProductPageContent);

ProductPageContent.propTypes = {
  classes: PropTypes.object.isRequired,
  currencyPrice: PropTypes.string.isRequired,
  singleProduct: PropTypes.object.isRequired,
  addToCart: PropTypes.func.isRequired
};
