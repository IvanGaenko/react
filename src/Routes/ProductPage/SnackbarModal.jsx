import React, { Component } from "react";
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Snackbar from '@material-ui/core/Snackbar';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';

const styles = () => ({
    close: {
      padding: 20,
    }
  });

class SnackbarModal extends Component {
  render() {
    const { classes, handleClose, open } = this.props;
    const { title, author } = this.props.singleProduct;
      return (
        <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'right',
        }}
        open={open}
        autoHideDuration={4000}
        onClose={handleClose}
        ContentProps={{
          'aria-describedby': 'message-id',
        }}
        message={
        <span id="message-id">
          <div>{title}</div>
          <div>{author}</div>
          <div>added to Cart</div>
        </span>
        }
        action={[
          <Button
            key="undo"
            color="inherit"
            size="small"
            onClick={handleClose}
          >
            Close
          </Button>,
          <IconButton
            key="close"
            aria-label="Close"
            color="inherit"
            className={classes.close}
            onClick={handleClose}
          ></IconButton>,
        ]}
      />
    )
  }
};

export default withStyles(styles)(SnackbarModal);

SnackbarModal.propTypes = {
  classes: PropTypes.object.isRequired,
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired
};
