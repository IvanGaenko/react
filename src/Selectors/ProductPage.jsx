import { createSelector } from "reselect";

const products = (state, props) => {
  const id = props.match.params.id;
  return { ProductList: [...state], id: Number(id) };
};

export const productSelector = createSelector(
  products,
  productsWithId =>
    productsWithId.ProductList.find(item => item.id === productsWithId.id)
);
