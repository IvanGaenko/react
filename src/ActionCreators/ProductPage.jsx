import * as constants from "../ActionTypes/ProductPage";
import { errorTrue, errorFalse } from './errorAction';

export const getSingleProductCreator = (products) => {
  return {
    type: constants.GET_SINGLE_ITEM,
    payload: 
      products
  }
};

export const getSingleProduct = id => dispatch => {
  let status;
  fetch(`http://localhost:3001/posts/${id}`)
    .then(responce => {
      status = responce.status;
      return responce.json();
    })
    .then(products => 
    {
      if (status >= 400 && status < 500) {
        errorTrue(dispatch, status);
      }
      dispatch(getSingleProductCreator(products));
      dispatch(errorFalse());
    }
    );
};
