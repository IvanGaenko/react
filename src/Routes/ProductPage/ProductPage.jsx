import React, { Component } from "react";
import { Link } from "react-router-dom";
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

import ProductPageContent from './ProductPageContent';
import SnackbarModal from './SnackbarModal';

const styles = () => (
  {
    root: {
      display: 'flex',
      alignItems: 'center',
      maxWidth: 1200,
      marginLeft: 'auto',
      marginRight: 'auto'
    },
    container: {
      maxWidth: '100%',
      marginLeft: 'auto',
      marginRight: 'auto',
      display: 'flex'
    }
  }
);

class ProductPage extends Component {
  state = {
    open: false
  };

  handleClick = () => {
    this.setState({ open: true });
  };

  handleClose = (event, reason) => {
    if (reason === 'clickaway') {
      return;
    }
    this.setState({ open: false });
  };
  
  componentDidMount() {
    const { getSingleProduct, match } = this.props;
    getSingleProduct(match.params.id);
  }

  addToCart = () => {
    const { addCart, singleProduct } = this.props;
    addCart(singleProduct);
    this.handleClick();
  }

  nextCart = () => {
    const { getNextProduct, singleProduct } = this.props;
    getNextProduct(singleProduct.id);
  }

  prevCart = () => {
    const { getPrevProduct, singleProduct } = this.props;
    getPrevProduct(singleProduct.id);
  }

  render() {
    const { classes, currencyArticle, currencyValue } = this.props;
    const { id, price } = this.props.singleProduct;

    const currencyPrice = `${currencyArticle + " " + (price * currencyValue)}`;
    return (
    <div className={classes.root}>
      <div className={classes.container}>
        <div>
          <Link to={`/posts/${id - 1}`}>
            <Button
              variant="contained"
              color="primary"
              onClick={this.prevCart}>
              Prev
            </Button>
          </Link>

          <Link to="/posts/">
            <Button
              variant="contained"
              color="primary">
              Back
            </Button>
          </Link>
        </div>

        <ProductPageContent
          currencyPrice={currencyPrice}
          singleProduct={this.props.singleProduct}
          addToCart={this.addToCart}
        />
      
        <Link to={`/posts/${id + 1}`}>
          <Button
            variant="contained"
            color="primary"
            onClick={this.nextCart}>
            Next
          </Button>
        </Link>
      </div>

      <SnackbarModal
        singleProduct={this.props.singleProduct}
        handleClose={this.handleClose}
        open={this.state.open}
      />
    </div>
  )}
};

export default withStyles(styles)(ProductPage);

ProductPage.propTypes = {
  classes: PropTypes.object.isRequired,
  singleProduct: PropTypes.object.isRequired,
  getSingleProduct: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired,
  addCart: PropTypes.func.isRequired
};
