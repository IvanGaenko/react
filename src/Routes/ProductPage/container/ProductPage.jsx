import { connect } from "react-redux";

import ProductPage from "../ProductPage";
import { getSingleProduct } from "../../../ActionCreators/ProductPage";
import { productSelector } from "../../../Selectors/ProductPage";

const mapStateToProps = (state, props) => ({
  singleProduct: productSelector(state.ProductList.products, props) || state.ProductPage,
  error: state.errorReducer
});

const mapDispatchToProps = dispatch => ({
  getSingleProduct: (products) => dispatch(getSingleProduct(products))
});

export default connect(mapStateToProps, mapDispatchToProps)(ProductPage);
