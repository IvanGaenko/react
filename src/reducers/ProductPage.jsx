import * as constants from "../ActionTypes/ProductPage";

const productDetails = (state = {}, action) => {
  const { type, payload } = action;
  switch (type) {
    
    case constants.GET_SINGLE_ITEM: 
      return {
        ...state,
        ...payload
      };

    case constants.GET_NEXT_ITEM: 
      return {
        ...state,
        ...payload
      };

    case constants.GET_PREV_ITEM: 
      return {
        ...state,
        ...payload
      };

    default:
      return state;
  }
};

export default productDetails;
