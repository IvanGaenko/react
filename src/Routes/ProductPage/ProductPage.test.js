import React from "react";
import { mount } from "enzyme";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import  ProductPage  from "./ProductPage";

describe("ProductPage", () => {
  const props = {
    singleProduct: {
    id: 1,
    title: "title",
    author: "author",
    img: "img",
    price: "100",
    year: "1920",
    },
    getSingleProduct: jest.fn(),
    addCart: jest.fn(),
    match: { params: { itemId: 1 } },
    myCurrency: {
      currentCurrency: [{
      "mark": "USD",
      "value": 1
    }]}
  };

  let component;
  beforeEach(() => {
    component = mount(
      <BrowserRouter>
        <ProductPage {...props} />
      </BrowserRouter>
    );
  });

  it("Should render correctly", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <ProductPage {...props} />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should renders product descriptions", () => {
    const typography = component.find("Typography");
    expect(typography.at(0).text()).toBe(props.singleProduct.title);
    expect(typography.at(1).text()).toBe(props.singleProduct.author);
    expect(typography.at(2).text()).toBe(`Year: ${props.singleProduct.year}`);
    expect(typography.at(3).text())
    .toBe(`Price: ${props.myCurrency.currentCurrency[0].mark + " " + props.singleProduct.price}`);
    expect(typography.children().length).toBe(4);
  });

  it("Should call function getSingleProduct in componentDidMount", () => {
    expect(props.getSingleProduct).toBeCalled();
  });

  it("Call function addToCart on click button 'Add to Cart'", () => {
    expect(props.addCart).not.toBeCalled();
    component.find("Button#add-cart").simulate("click");
    expect(props.addCart).toBeCalled();
  });
});
