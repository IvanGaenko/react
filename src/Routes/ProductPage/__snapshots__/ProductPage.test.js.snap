// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPage Should render correctly 1`] = `
<div
  className="ProductPage-root-1"
>
  <div
    className="ProductPage-container-2"
  >
    <div>
      <a
        href="/posts/0"
        onClick={[Function]}
      >
        <button
          className="MuiButtonBase-root-29 MuiButton-root-3 MuiButton-contained-14 MuiButton-containedPrimary-15 MuiButton-raised-17 MuiButton-raisedPrimary-18"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-4"
          >
            Prev
          </span>
          <span
            className="MuiTouchRipple-root-112"
          />
        </button>
      </a>
      <a
        href="/posts/"
        onClick={[Function]}
      >
        <button
          className="MuiButtonBase-root-29 MuiButton-root-3 MuiButton-contained-14 MuiButton-containedPrimary-15 MuiButton-raised-17 MuiButton-raisedPrimary-18"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-4"
          >
            Back
          </span>
          <span
            className="MuiTouchRipple-root-112"
          />
        </button>
      </a>
    </div>
    <div
      className="MuiPaper-root-36 MuiPaper-elevation1-39 MuiPaper-rounded-37 MuiCard-root-35 ProductPageContent-card-32"
    >
      <div
        className="MuiCardMedia-root-63 ProductPageContent-media-33"
        title={1}
      />
      <div
        className="MuiCardContent-root-65"
      >
        <h2
          className="MuiTypography-root-66 MuiTypography-headline-71 MuiTypography-gutterBottom-93"
        >
          title
        </h2>
        <h2
          className="MuiTypography-root-66 MuiTypography-headline-71 MuiTypography-gutterBottom-93"
        >
          author
        </h2>
        <p
          className="MuiTypography-root-66 MuiTypography-body1-75"
        >
          Year: 
          1920
        </p>
        <p
          className="MuiTypography-root-66 MuiTypography-body1-75"
        >
          Price: 
          UAH 100
        </p>
      </div>
      <div
        className="MuiCardActions-root-101"
      >
        <button
          className="MuiButtonBase-root-29 MuiButton-root-3 MuiButton-text-5 MuiButton-textPrimary-6 MuiButton-flat-8 MuiButton-flatPrimary-9 MuiButton-sizeSmall-26 ProductPageContent-button-34 MuiCardActions-action-103"
          disabled={false}
          id="add-cart"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-4"
          >
            Buy
          </span>
          <span
            className="MuiTouchRipple-root-112"
          />
        </button>
      </div>
    </div>
    <a
      href="/posts/2"
      onClick={[Function]}
    >
      <button
        className="MuiButtonBase-root-29 MuiButton-root-3 MuiButton-contained-14 MuiButton-containedPrimary-15 MuiButton-raised-17 MuiButton-raisedPrimary-18"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-4"
        >
          Next
        </span>
        <span
          className="MuiTouchRipple-root-112"
        />
      </button>
    </a>
  </div>
</div>
`;
